Feat/7 loaction

rabitis99 opened the pull request:
## 📌 관련 이슈
- Closes #7  <!-- 연결된 이슈 번호 기재 -->

---

## ✨ 기능 요약
<!--
| 항목 | 내용 |
|------|------|
| 🆕 기능명 | 기능 이름을 간단히 요약 |
| 🔍 목적 | 해당 기능이 왜 필요한지 간략 설명 |
| 🛠️ 변경사항 | 핵심 변경 포인트 간략히 요약 (예: UI 추가, API 연동 등) |
-->
1. 🔍 목적 | 사용자 위치 정보 저장(redis 이름 수정 여부)
2. 🔍 목적 | 근처 대피소 3개 알려주기
3. 🔍 목적 | 재난 사건 위치 전달하기
4. 🛠️ 변경사항 | 위치정보는 Double로 변경
---

## 📝 상세 내역
<!--
| 번호 | 내용 |
|------|------|
| 1️⃣ | 주요 로직/컴포넌트/서비스 등 변경 설명 |
| 2️⃣ | 관련 유틸, 공통 로직 수정 여부 |
| 3️⃣ | 리팩토링 또는 제거된 불필요 코드 |
-->

---

## ✅ 테스트 체크리스트
<!-- 테스트 할 내용이 있다면 작성해주세요 -->
- [ ] 기능 정상 작동 확인
- [ ] 예외/엣지 케이스 확인
- [ ] UI/UX 흐름 확인 (필요 시 캡처 또는 영상 첨부)
- [ ] 테스트 코드 작성 완료
- [ ] API 연동 확인 (요청/응답 정상 동작)

---

## 📸 포스트맨 캡처 사진
<!-- 필요한 경우 Postman 캡처나 Swagger 스크린샷 첨부 -->

---

## 🙋‍♀️ 리뷰어 참고사항
<!-- 
- 테스트 방법
- 주요 로직 설명
- 리뷰 시 중점적으로 봐주셨으면 하는 부분
- 코드 스타일 관련 사항 등
-->
1. userDetail,  report repository 확인 후 코드 수정 예정
2. 1번 수정 후, 프론트 수정예정
3. 테스트 결과 정상 작동 시, 백엔드,프론트엔드 각각 새로운 issue로 pr 예정

rabitis99 committed:
feat 위치관련 controller/service/redis

rabitis99 committed:
fix 위도 경도 Double로 변경

rabitis99 committed:
fix 코드 의존도 분리, 역할에 맞게 dto 나 utils 생성

MJK-One commented on src/main/java/com/example/emergencyassistb4b4/location/redis/LocationService.java:
```diff
@@ -0,0 +1,29 @@
+        Map<String, Double> coordinates = new HashMap<>();
+        coordinates.put("latitude", latitude);
+        coordinates.put("longitude", longitude);
+        redisTemplate.opsForHash().putAll("coordinates:" + userId, coordinates);
```
Geospatial 자료구조가 위치 기반 관련해서 더 좋다고 알고 있는데 혹시 hash 자료구조를 사용한 이유가 있을까요??

MJK-One replied:
위도/경도를 Hash 자료구조에 저장하셨는 데, 혹시 hash를 사용한 이유가 있을까요? 특정 위치 반경 내 사용자 탐색이나 거리 계산이 필요할 경우 Geospatial이 더 효율적인 것으로 알고 있어서 지금 상황에서 hash로 충분한지 궁금합니다!!

rabitis99 committed:
fix ìfix Geospatia 도입

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/global/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,49 @@
+    @Bean
+    public RedisTemplate<String, Object> objectRedisTemplate() {
+        RedisTemplate<String, Object> template = new RedisTemplate<>();
+        template.setConnectionFactory(redisConnectionFactory());
```
아래와 동일한 형태로 `connectionFactory`를 주입해주면 더 좋을 것 같아요!

```
public RedisTemplate<String, Object> objectRedisTemplate(RedisConnectionFactory connectionFactory) {
     RedisTemplate<String, Object> template = new RedisTemplate<>();
     template.setConnectionFactory(connectionFactory);
     // ...
}
```

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/global/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,49 @@
+        template.setKeySerializer(new StringRedisSerializer());
+        template.setHashKeySerializer(new StringRedisSerializer());
+        // 필요한 경우 ValueSerializer 지정 (예: JdkSerializationRedisSerializer, Jackson2JsonRedisSerializer 등)
+        template.afterPropertiesSet();
```
Spring이 `@Bean` 등록 시 자동으로 호출해주기 때문에 `afterPropertiesSet()`는 생략 가능하다고 해요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/location/controller/KakaoMapController.java:
```diff
@@ -0,0 +1,59 @@
+import static com.example.emergencyassistb4b4.global.status.SuccessStatus.SHELTER_SEARCH_SUCCESS;
+
+@RestController
+@RequestMapping()
```
공통 경로가 없다면 이 부분은 생략해도 좋을 것 같습니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/location/controller/KakaoMapController.java:
```diff
@@ -0,0 +1,59 @@
+            @RequestParam(defaultValue = "1000") int radiusMeter,
+            @RequestParam(defaultValue = "3600") long secondsAgo) {
+
+        Duration duration = Duration.ofSeconds(secondsAgo);
```
이 부분은 Service 내부로 옮기는 게 더 좋을 것 같아요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/location/controller/LocationController.java:
```diff
@@ -0,0 +1,42 @@
+    @PostMapping("/region")
+    public ResponseEntity<ApiResponse<String>> saveRegion(@RequestBody RegionRequestDto dto) {
+        locationService.saveRegion(dto.getUserId(), dto.getSi(), dto.getGu());
+        return ApiResponse.onSuccess(LOCATION_SAVE_SUCCESS,"시/구 정보 저장 완료");
```
`ApiResponse` 의 `payload` 부분은 실제 비즈니스 데이터를 담는 용도로만 사용해 주시는게 좋을 것 같아요!
비즈니스 데이터가 없을 경우에는 `return ApiResponse.onSuccess(LOCATION_SAVE_SUCCESS, null);`로 작성해주시고,
안내 메시지는 `SuccessStatus`의 `message` 부분에만 작성해주세요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/location/controller/LocationController.java:
```diff
@@ -0,0 +1,42 @@
+//            throw new ForbiddenException(ErrorStatus.USER_NOT_VOLUNTEER);
+//        }
+        locationService.saveCoordinates(dto.getUserId(), dto.getLatitude(), dto.getLongitude());
+        return ApiResponse.onSuccess(LOCATION_SAVE_SUCCESS,"위치 좌표 저장 완료");
```
이 부분도 함께 수정 부탁드립니다!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/location/redis/LocationService.java:
```diff
@@ -0,0 +1,84 @@
+    // 행정 구역 저장
+    public void saveRegion(Long userId, String si, String gu) {
+        String regionKey = si + " " + gu;
+        redisTemplate.opsForValue().set(regionKey, userId.toString());
```
`opsForValue()` 부분을 List 나 Set 타입으로 변경해야 할 것 같아요!

khg9900 commented on src/main/java/com/example/emergencyassistb4b4/location/redis/LocationService.java:
```diff
@@ -0,0 +1,84 @@
+
+    // 재난 알림 사용시(si + " " + gu 형태)
+    public List<String> getRegion(String region) {
+        return redisTemplate.opsForList().range(region, 0, -1);
```
`saveRegion` 수정 후 동일하게 맞춰주시면 될 것 같습니다!

rabitis99 committed:
fix pr 피드백 및 redis repository와 service로 변경
-redis config 수정
-불필요한 @RequestMapping()
-Duration withinTime 서비스단으로
-"시/구 정보 저장 완료","위치 좌표 저장 완료" null로 변경

khg9900 reviewed (approved):
고생많으셨습니다☺️👍

Prototype-Km reviewed (approved):
고생하셨습니다

MJK-One reviewed (approved):
고생하셨습니다!

rabitis99 committed:
Merge branch 'dev/backend' into feat/7-loaction
